src/devices/SenseHat/SenseHatLedMatrixI2c.cs

```diff
@@ -0,0 +1,145 @@
+        private const int BOffset = 16;
+        private I2cDevice _i2c;
+
+        public SenseHatLedMatrixI2c(I2cDevice i2cDevice = null)
+        {
+            _i2c = i2cDevice ?? CreateDefaultI2cDevice();
+            Clear(Color.Black);
+        }
+
+        public override void Write(ReadOnlySpan<Color> colors)
+        {
+            if (colors.Length != NumberOfPixels)
+                throw new ArgumentException($"`{nameof(colors)}` must have exactly {NumberOfPixels} elements.");
+
+            Span<byte> buffer = stackalloc byte[FrameBufferLength + 1];
+
+            // Register address of first pixel
+            buffer[0] = 0;
+
+            Span<byte> frameBuffer = buffer.Slice(1);
+
+            for (int i = 0; i < NumberOfPixels; i++)
+            {
+                (int x, int y) = IndexToPosition(i);
+                int address = PositionToAddress(x, y);
+                (byte r, byte g, byte b) = DestructColor(colors[i]);
+                frameBuffer[address + ROffset] = r;
+                frameBuffer[address + GOffset] = g;
+                frameBuffer[address + BOffset] = b;
+            }
+
+            _i2c.Write(buffer);
+        }
+
+        public override void Clear(Color color = default(Color))
+        {
+            Span<byte> buffer = stackalloc byte[FrameBufferLength + 1];
+
+            // Register address of first pixel
+            buffer[0] = 0;
+
+            Span<byte> frameBuffer = buffer.Slice(1);
+            const int channelLength = NumberOfPixelsPerRow;
+            (byte r, byte g, byte b) = DestructColor(color);
+
+            for (int y = 0; y < NumberOfPixelsPerColumn; y++)
+            {
+                for (int x = 0; x < NumberOfPixelsPerRow; x++)
+                {
+                    frameBuffer[x] = r;
+                }
+
+                frameBuffer = frameBuffer.Slice(channelLength);
+
+                for (int x = 0; x < NumberOfPixelsPerRow; x++)
+                {
+                    frameBuffer[x] = g;
+                }
+
+                frameBuffer = frameBuffer.Slice(channelLength);
+
+                for (int x = 0; x < NumberOfPixelsPerRow; x++)
+                {
+                    frameBuffer[x] = b;
+                }
+
+                frameBuffer = frameBuffer.Slice(channelLength);
+            }
+
+            _i2c.Write(buffer);
+        }
+
+        public override void SetPixel(int x, int y, Color color)
+        {
+            (byte r, byte g, byte b) = DestructColor(color);
+
+            int address = PositionToAddress(x, y);
+            Span<byte> encoded = stackalloc byte[2];
+
+            encoded[0] = (byte)(address + ROffset);
+            encoded[1] = r;
+            _i2c.Write(encoded);
+
+            encoded[0] = (byte)(address + GOffset);
+            encoded[1] = g;
+            _i2c.Write(encoded);
+
+            encoded[0] = (byte)(address + BOffset);
+            encoded[1] = b;
+            _i2c.Write(encoded);
```

Comment:
Is it necessary to send 3 separate writes? Would sending a single 6-byte write work?

Reply:
yes, those are 3 separate transactions (I2C numbers the data it sends) - I actually spent quite a bit of time debugging why writing 6 bytes at once doesn't do what I want :smile:

Reply:
> I actually spent quite a bit of time debugging why writing 6 bytes at once doesn't do what I want

Care to share your findings? What does it end up doing instead if you write at once? Do you know why it doesn't work?